p2p/exchange_metrics.go

```diff
@@ -0,0 +1,165 @@
+	blockedPeersNum     metric.Int64Counter
+
+	trackerPeersNum     atomic.Int64
+	trackedPeersNumInst metric.Int64ObservableGauge
+	trackedPeersNumReg  metric.Registration
+
+	disconnectedPeersNum     atomic.Int64
+	disconnectedPeersNumInst metric.Int64ObservableGauge
+	disconnectedPeersNumReg  metric.Registration
+}
+
+func newExchangeMetrics() (m *exchangeMetrics, err error) {
+	m = new(exchangeMetrics)
+	m.headRequestTimeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_head_time_hist",
+		metric.WithDescription("exchange client head request time in milliseconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseSizeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_resp_size_hist",
+		metric.WithDescription("exchange client header response size in bytes"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.responseTimeInst, err = meter.Int64Histogram(
+		"hdr_p2p_exch_clnt_resp_time_hist",
+		metric.WithDescription("exchange client response time in milliseconds"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_trck_peer_num_gauge",
+		metric.WithDescription("exchange client tracked peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.trackedPeersNumReg, err = meter.RegisterCallback(m.observeTrackedPeers, m.trackedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_exch_clnt_disc_peer_num_gauge",
+		metric.WithDescription("exchange client tracked disconnected peers number"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.disconnectedPeersNumReg, err = meter.RegisterCallback(m.observeDisconnectedPeers, m.disconnectedPeersNumInst)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *exchangeMetrics) head(ctx context.Context, duration time.Duration, headersReceived int, tp, status string) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.headRequestTimeInst.Record(ctx,
+			duration.Milliseconds(),
+			metric.WithAttributes(
+				attribute.Int(headerReceivedKey, headersReceived),
+				attribute.String(headTypeKey, tp),
+				attribute.String(headStatusKey, status),
+			),
+		)
+	})
+}
+
+func (m *exchangeMetrics) response(ctx context.Context, size uint64, duration time.Duration, err error) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.responseSizeInst.Record(ctx,
+			int64(size),
+			metric.WithAttributes(attribute.Bool(failedKey, err != nil)),
+		)
+		m.responseTimeInst.Record(ctx,
+			duration.Milliseconds(),
+			metric.WithAttributes(attribute.Bool(failedKey, err != nil)),
+		)
+	})
+}
+
+func (m *exchangeMetrics) peersTracked(num int) {
+	m.observe(context.Background(), func(context.Context) {
+		m.trackerPeersNum.Add(int64(num))
+	})
+}
+
+func (m *exchangeMetrics) peersDisconnected(num int) {
+	m.observe(context.Background(), func(context.Context) {
+		m.disconnectedPeersNum.Add(int64(num))
+	})
+}
```

Comment:
I'm not sure about this metric tbh, because it shows the total amount of disconnected peers over time. It could be the same peer connecting/disconnecting several times.

Reply:
> I'm not sure about this metric tbh, because it shows the total amount of disconnected peers over time. 

It shows the number of disconnected peers at the moment and not over time. But also depends on what you mean by over time. 

> It could be the same peer connecting/disconnecting several times.

IMO, that's fine. If goes from disconnected -> connected and vice versa, it will be reflected in both cases. 
My motivation for this metric is to give some visibility on the internal state of the disconnected map, so we can see a burst of disconnections or leaks.

Reply:
Okay i don't mind

Reply:
You are decreasing the amount of connected peers each time during peer is disconnected and increasing amount of disconnected. So, it will end up with eg 100 connected peers and 1000 disconnected after one month. DisconnectedPeers shows how many peers were disconnected after system booted up(overall) and ConnectedPeers shows how many connected peers we have right now. I was trying to say that DisconnectedPeers does not show smth useful from the user's perspective.

Reply:
`p.metrics.peersDisconnected(-deletedDisconnectedNum)`

Did not notice that `-`. No more questions